scrapy/extensions/throttle.py

```diff
@@ -91,3 +112,193 @@ def _adjust_delay(self, slot, latency, response):
+        #                           + (60 seconds - current second))
+        # if interval is day, return:  get_seconds(1 day) - (get_seconds(1 hour) * (24 hours - current hour) +
+        #                          (get_seconds(1 min) * (60 minutes - current_minutes) + (60 seconds - current second))
+        spent_time = 0
+        if interval == TimeInterval.SECOND:
+            return self.get_seconds(interval) - spent_time
+        spent_time += time_stamp.second
+        if interval == TimeInterval.MINUTE:
+            return self.get_seconds(interval) - spent_time
+        spent_time += time_stamp.minute * self.get_seconds(TimeInterval.MINUTE)
+        if interval == TimeInterval.HOUR:
+            return self.get_seconds(interval) - spent_time
+        spent_time += time_stamp.hour * self.get_seconds(TimeInterval.HOUR)
+        return self.get_seconds(interval) - spent_time
+
+    @staticmethod
+    def get_timestamp_key(t: datetime, degree: TimeInterval):
+        return t.strftime(degree.value[0])
+
+    @staticmethod
+    def get_delta(degree: TimeInterval):
+        return timedelta(seconds=TimeIntervalManager.get_seconds(degree))
+
+    @staticmethod
+    def get_parts(degree: TimeInterval):
+        return degree.value[2]
+
+    @staticmethod
+    def get_seconds(degree: TimeInterval):
+        return degree.value[1]
+
+
+class SlidingWindow:
+
+    """
+    This structure implements keeps track of the number of requests over the previous and current time intervals.
+    """
+
+    def __init__(self, interval: TimeInterval, current: datetime, current_count=1, previous_count=0):
+        self.interval = interval
+        self.current = current
+        self.current_count = current_count
+        self.previous_count = previous_count
+        self._tim = TimeIntervalManager()
+
+    def add_request(self, sent_at: datetime):
+        label = self.get_key(sent_at)
+        if label != self.get_key(self.current):
+            # slide window
+            # case 1. next window. case 2, previous window was blank. (no requests in the last interval).
+            if self.is_next_window(label):
+                self.reset_window(sent_at, self.current_count)
+            else:
+                self.reset_window(sent_at, 0)
+        else:
+            self.current_count += 1
+
+    def reset_window(self, new_current, new_previous_count, set_blank=False):
+        self.current = new_current
+        self.current_count = 0 if set_blank else 1
+        self.previous_count = new_previous_count
+
+    def is_next_window(self, label: str):
+        return self.get_key(self.current + self._tim.get_delta(self.interval)) == label
+
+    def get_key(self, _datetime):
+        return self._tim.get_timestamp_key(_datetime, self.interval)
+
+    def __repr__(self):
+        return "SlidingWindow: {}\n Current Window: {}\n Current Count{}\n Previous Count{}" \
```

Comment:
A `__repr__` with newlines sounds strange to me, but that might be subjective.
However, we're trying to avoid backslashes, so please modify this to something like
```python
return (
    "stuff {} more stuff {} ..."
).format(
    var1,
    var2,
    ...
)
```

or maybe

```python
msg = "stuff {} more stuff {} ..."
return msg.format(
    var1,
    var2,
    ...
)
```